core/miner/storage_fsm/impl/basic_precommit_policy.cpp

```diff
@@ -39,17 +40,34 @@ namespace fc::mining {
       end = epoch + duration_;
     }
 
-    *end += kWPoStProvingPeriod - (*end % kWPoStProvingPeriod)
-            + proving_boundary_ - 1;
+    const auto min_exp = epoch + kMinSectorExpiration + kWPoStProvingPeriod;
+
+    if (end < min_exp) {
+      end = min_exp;
+    }
 
     return *end;
   }
 
-  BasicPreCommitPolicy::BasicPreCommitPolicy(std::shared_ptr<FullNodeApi> api,
-                                             ChainEpoch proving_boundary,
-                                             ChainEpoch duration)
+  BasicPreCommitPolicy::BasicPreCommitPolicy(
+      std::shared_ptr<FullNodeApi> api,
+      ChainEpoch proving_boundary,
+      std::chrono::seconds sector_lifetime)
       : api_(std::move(api)),
         proving_boundary_(proving_boundary),
-        duration_(duration),
-        logger_(common::createLogger("basic pre commit policy")) {}
+        logger_(common::createLogger("basic pre commit policy")) {
+    ChainEpoch sector_lifetime_ =
+        std::ceil(sector_lifetime.count() / double(kEpochDurationSeconds));
```

Comment:
or is it `floor`?

https://github.com/filecoin-project/lotus/blob/master/extern/storage-sealing/precommit_policy.go#L110